README.md
One Medical PCR sniper
======================

This application books a COVID PCR test appointment with One Medical once it becomes available.

If you can book an appointment directly from One Medical site or with another provider, you should do so. Otherwise, you can use this application so that it

1. Drives a web browser to log into One Medical account,
2. Searches for available appointments, and
3. Books one matching your criteria.

It keeps searching until an appointment matching your criteria becomes available.

This program does exactly what One Medical suggests you should do. It "checks back regularly" to see if appointments became available. Here is that the One Medical site says, literally:

![](data/check_back.png)

To book an appointment, you will have to

1. Set up Haskell Stack and compile the application.
2. Download and run [WebDriver for Chrome](https://chromedriver.chromium.org/downloads) matching the version of Chrome installed at your system.
3. Configure your credentials and appointment search criteria.
4. Dry-run the program to confirm it correctly searches for appointments.
5. Run the program for an extended period until the desired appointment is available.

I suggest you start with loose criteria for step 3, get an appointment, and then refine your search criteria in step 3 and repeat from step 4. You should cancel an existing appointment as soon as possible once you get a better one (closer or earlier). Click on a menu in the top right corner of the site and select "Appointments." You will see your appointments there and can cancel the less-matching one.

![](data/appointments.png)

You may want to write a function in Haskell to customize your appointment selection criteria.

Setup Haskell Stack and compile the application
-----------------------------------------------

On Linux/macOS, run

`curl -sSL https://get.haskellstack.org/ | sh`

On Windows, use [this installer](https://get.haskellstack.org/stable/windows-x86_64-installer.exe).

Read [Stack Intall/Upgrade](https://docs.haskellstack.org/en/stable/install_and_upgrade/) instructions for more details.

Once you get Stack installed, you should go to the directory you cloned this repo to and build the application:

```
$ stack build
```

The build will take about 20 minutes to complete.

You will see something like the following at the end:

```
Installing executable one_medical in /Users/ilya/safeplace/projects/projects_log/20220105_pcr_sniper/.stack-work/install/x86_64-osx/bb55b0183003d9fd6f944a319ba19b13f05e132815290cdc419812323b8f3d76/8.10.7/bin
Registering library for pcr-sniper-0.0.1.0..
```

Try running the application as follows. It should fail.

```
$ stack run
one_medical: HttpExceptionRequest Request {
  host                 = "localhost"
  port                 = 9515
  secure               = False
  requestHeaders       = [("Accept","application/json;charset=UTF-8"),("Content-Type","application/json;charset=UTF-8")]
  path                 = "/wd/hub/session"
  queryString          = ""
  method               = "POST"
  proxy                = Nothing
  rawBody              = False
  redirectCount        = 10
  responseTimeout      = ResponseTimeoutDefault
  requestVersion       = HTTP/1.1
}
 (ConnectionFailure Network.Socket.connect: <socket: 49>: does not exist (Connection refused))
```

This failure is expected because we didn't install and run WebDriver for Chrome yet.

Download and run WebDriver for Chrome
-------------------------------------

Open Chrome and type `chrome://settings/help` in the address bar. You will see something like the following:

![](data/chrome_version.png)

The page will have a Chrome version number. I highlighted it in red. Go to the WebDriver for Chrome [Downloads](https://chromedriver.chromium.org/downloads) and get the ChromeDriver of the same version as your Chrome. I would get the highlighted version to match my version of Chrome. Your version may be different:

![](data/chrome_driver.png)

You will need to pick an archive for Linux, Mac, Mac M1, or Windows.

Run the Chrome Driver with the following arguments once downloaded and extracted:

```
$ ./chromedriver --port=9515 --url-base=/wd/hub
Starting ChromeDriver 97.0.4692.71 (adefa7837d02a07a604c1e6eff0b3a09422ab88d-refs/branch-heads/4692@{#1247}) on port 9515
Only local connections are allowed.
Please see https://chromedriver.chromium.org/security-considerations for suggestions on keeping ChromeDriver safe.
ChromeDriver was started successfully.
```

Configure your credentials and appointment search criteria
----------------------------------------------------------

Open `app/one_medical.hs` in an editor.

You will see the following declaration for application parameters:

```haskell
params
  = FindAppointmentParams
  { user = "<your email registered with OneMedical>"
  , password = "<your OneMedical password>"
  , area = "SF Bay Area"
  , appointmentType = "COVID-19 PCR Test"
  , lastDay = fromGregorian 2022 01 20
  , predicate = locationsInP southBayLocations
  }
```

Provide your One Medical email and password as values of `user` and `password` fields of the `FindAppointmentParams` record. For example:

```haskell
params
  = FindAppointmentParams
  { user = "someone@yahoo.com"
  , password = "MySecretPass1"
  [...]
  }
```

You can change `area` to wherever region where One Medical is available, i.e., "New York," "Chicago," etc.

Specify `lastDay`. The program will search appointments from today until the value of the `lastDay`. `lastDay = fromGregorian 2022 01 20` means that it will search from today until January 20th, 2022.

Configure appointment search criteria. An obvious use case is to provide locations of interest. The simplest way is to feed a list of locations into the `locationsInP` function. For example, the following will only look for appointments in three offices.

```haskell
predicate = locationsInP ["Redwood City", "Palo Alto", "Sunnyvale"]
```

Location strings should match exactly "Select Locations" in One Medical UI. For example, "Sunnyvale" is one of the valid strings:

![](data/select_locations.png)

An example list of "south bay" locations is available in `app/one_medical.hs`, `predicate = locationsInP southBayLocations` where:

```haskell
southBayLocations =
  [ "Sunnyvale"
  , "Sunnyvale Outdoor Testing Site"
  , "San Mateo - Bay Meadows"
  , "San Jose - The Alameda"
  , "San Jose - North First"
  , "Redwood City"
  , "Phillips Brooks School COVID Testing Site" -- Menlo Park
  , "Palo Alto"
  ]
```

You may define your own list.

You may search for appointments "not in locations", for example:

```haskell
params
  = FindAppointmentParams
  { ...
  , predicate = not . locationsInP [ "UCSF One Medical Testing Site"
                                   , "Sunnyvale Outdoor Testing Site"
                                   ]
  }
```

Predicate function takes an `Appointment` and must return `True` if the appointment matches your criteria. You may write a custom `predicate` function in Haskell. The following example accepts an appointment if the provider's location doesn't have "San Jose" infix and the time of an appointment is after 10 am:

```haskell
params
  = FindAppointmentParams
  { ...
  , predicate = \appt ->
      not ("San Jose" `T.isInfixOf` location (providerInfo appt))
      && (localTimeOfDay (time appt) >= TimeOfDay 10 0 0)
  }
```

Dry-run the program to confirm it correctly searches for appointments
---------------------------------------------------------------------

If an appointment you need is available, just book it. You don't need this program. Otherwise, you should run this program in the "dry-run mode" first to see if it can NOT find an appointment you need and loops. This step will guarantee that you don't book an appointment you don't need.

Now that you have your Chrome Web Driver running, credentials, and search criteria configured, you can check that the program keeps scanning for appointments:

1. Make sure you have the following definition of `main` in `app/one_medical.hs`

    ```haskell
    main = runSession remoteConfig findAppointments
    ```
    
    We don't want it to be `bookAppointment` yet.
    
2. Execute `stack run`
3. You should observe something like this:

    ![](data/sniper.gif)
    
The program goes into One Medical account and loops over the appointment schedule because it can't find a matching appointment. It prints appointments it considered to the console. You can examine that these appointments are indeed NOT what you need. Example output:
    
```
No appointments found between 2022-01-14 and 2022-01-16
No appointments match predicate:
[ Appointment
    { providerInfo = ProviderInfo
        { location = "San Mateo - Bay Meadows" }
    , time = 2022 - 01 - 19 10 : 50 : 00
    , element = Element "0.37229807800837955-298"
    }
, Appointment
    { providerInfo = ProviderInfo
        { location = "Burlingame" }
    , time = 2022 - 01 - 19 14 : 30 : 00
    , element = Element "0.37229807800837955-299"
    }
...
]
pageLastDay >= lastDay: 2022-01-19 >= 2022-01-17
```

Appointments in this sample got rejected because `lastDay = fromGregorian 2022 01 17`. The requested last day is January 17th, but the nearest appointments are on January 19th.

If the program stops at this stage, then you've found an appointment that matches your criteria. Just book it manually.

Run the program until the desired appointment is available
--------------------------------------------------------

Now that you know that the program doesn't find a matching appointment, you should leave it running in a "booking" mode so that it books an appointment matching your criteria as soon as it becomes available.

Change the `app/one_medical.hs` such that

```haskell
main = runSession remoteConfig bookAppointment
```

Execute `stack run` and wait until the program stops after a successful booking. The program is not 100% reliable. It may fail for a few reasons. So, you may want to check on it once in a while to see if it still runs.



ChangeLog.md
# Changelog for pcr-sniper

## 0.0.1.0

Initial version. Searches for and books a One Medical PCR COVID test appointment.


OneMedical.hs
{-# LANGUAGE DuplicateRecordFields #-}
{-# LANGUAGE OverloadedStrings     #-}
{-# LANGUAGE ScopedTypeVariables   #-}

module OneMedical
  ( findAppointmentsScenario
  , bookAppointmentScenario
  , FindAppointmentParams(..)
  , Appointment(..)
  , ProviderInfo(..)
  -- :Predicates:
  , allP
  , dayEndP
  , dayStartP
  , locationsInP
  , locationsP
  , timeStartP
  , timeEndP
  -- :Selenium:
  , remoteConfig
  , returnSession
  ) where

import           Control.Monad
import           Control.Monad.Catch
import           Control.Monad.IO.Class
import           Data.Maybe
import           Data.Text              (Text)
import qualified Data.Text              as T
import qualified Data.Text.Lazy         as LT
import           Data.Time              (Day, LocalTime (LocalTime), ParseTime,
                                         TimeOfDay (TimeOfDay), addDays,
                                         defaultTimeLocale, fromGregorian,
                                         parseTimeM)
import           Test.WebDriver
import           Test.WebDriver.Session
import           Text.Pretty.Simple

-- ./chromedriver --port=9515 --log-level=ALL --url-base=/wd/hub
-- ./chromedriver --port=9515  --url-base=/wd/hub
remoteConfig = useBrowser chrome defaultConfig { wdHost = "localhost"
                                               , wdPort = 9515
                                               }

data HomePage = HomePage
data ChooseAppointmentPage = ChooseAppointmentPage

defaultWait = 10000

login :: Text -> Text -> WD HomePage
login user pwd = do
  openPage "https://onemedical.com"
  click =<< findElem (ByPartialLinkText "Log in")
  loop
  where loop = do
          sendKeys user =<< findElem (ById "email")
          sendKeys pwd =<< findElem (ById "password")
          login <- findElem (ById "btn-login")
          loginDisplayed <- isDisplayed login
          logWD ("'Log In' button is displayed: " ++ show loginDisplayed)
          click login
          catch (
            findElem (ByPartialLinkText "Get Care")
            >> return HomePage
            ) $ \(e::SomeException) -> do
              logWD $ "Can't login, trying again"
              clearInput =<< findElem (ById "email")
              clearInput =<< findElem (ById "password")
              loop

gotoAppointmentSelection :: Text -> HomePage -> WD ChooseAppointmentPage
gotoAppointmentSelection location _  = do
  click =<< findElem (ByPartialLinkText "Get Care")
  sendKeys "COVID PCR Test"
    =<< findElem (ByXPath "//*[@formcontrolname='reasonForVisit']")
  select <- findElem (ByXPath "//*[@formcontrolname='serviceArea']")
  click =<< head <$> (selectItem select location)
  click =<< findElem (ByXPath "//*[@aria-label='Submit']")
  return ChooseAppointmentPage

selectAppointment :: Text -> ChooseAppointmentPage -> WD ()
selectAppointment appointmentType _ = do
  click
    =<< fmap head . filterM (fmap (T.isInfixOf appointmentType) . getText)
    =<< findElems (ByTag "h5")

logWD = liftIO . putStrLn

findFirstAppointments :: Day -> WD [Appointment]
findFirstAppointments lastDay = do
  exist <- checkProvidersExistOnPage
  if exist
    then do
      pageFirstDay <- toLastDay =<< findFirstDayElem
      pageLastDay  <- (toLastDay =<< findLastDayElem :: WD Day)
      if pageFirstDay > lastDay
        then do
          logWD
            $ "pageFirstDay > lastDay: "
            ++ show pageFirstDay
            ++ " > "
            ++ show lastDay
          return []
        else do
          appts <- readAppointments
          if null appts
            then do
              logWD
                $  "No appointments found between "
                ++ show pageFirstDay
                ++ " and "
                ++ show pageLastDay
              if pageLastDay >= lastDay
                then do
                  logWD
                    $ "pageLastDay >= lastDay: "
                    ++ show pageLastDay
                    ++ " >= "
                    ++ show lastDay
                  return []
                else do
                  nextAppointmentsPage
                  findFirstAppointments lastDay
            else
              return appts
    else do
      logWD "No providers found"
      return []

findAppointments :: Day -> (Appointment -> Bool) -> WD [Appointment]
findAppointments lastDay p = do
  appts <- findFirstAppointments lastDay
  if null appts
    then return []
    else do
      let appts'
            = filter
              (allP [ (<= nextDayStartTime lastDay) . time, p ])
              appts
      if not . null $ appts'
        then return appts'
        else do
          logWD
            $ "No appointments match predicate:\n"
            ++ (LT.unpack . pShowNoColor $ appts)
          pageLastDay  <- (toLastDay =<< findLastDayElem :: WD Day)
          if pageLastDay >= lastDay
            then do
              logWD
                $ "pageLastDay >= lastDay: "
                ++ show pageLastDay
                ++ " >= "
                ++ show lastDay
              return []
            else do
              nextAppointmentsPage
              findAppointments lastDay p

nextAppointmentsPage = click =<< findElem (ByXPath "//div/om-arrow-right/div")

findLastDayElem :: WD Element
findLastDayElem = findElem (ByXPath "//div[om-arrow-right]/div/div[3]")

findFirstDayElem :: WD Element
findFirstDayElem = findElem (ByXPath "//div[om-arrow-right]/div/div[1]")

toLastDay :: ParseTime t => Element -> WD t
toLastDay = fmap parseLastDay . getText

-- parseLastDay "Sat Jan 08" -> 2022-01-08
parseLastDay :: ParseTime t => Text -> t
parseLastDay
  = fromJust
  . parseTimeM True defaultTimeLocale "%Y %a %b %d"
  . ("2022 " ++ ) . T.unpack

data Appointment
  = Appointment
  { providerInfo :: ProviderInfo
  , time         :: LocalTime
  , element      :: Element
  } deriving Show

data ProviderInfo
  = ProviderInfo
  { location :: Text
  } deriving Show

data Provider
  = Provider
  { providerInfo :: ProviderInfo
  , appointments :: [Appointment]
  } deriving Show

readAppointments :: WD [Appointment]
readAppointments = do
  exist <- checkAppointmentsExistOnPage
  if exist
    then do
      setImplicitWait 0
      providers <- readProviders
      setImplicitWait defaultWait
      return . concat . map appointments $ providers
    else
      return []

waitAppointmentsPageLoaded = do
  elems <- findElems (ByPartialLinkText "Back to Appointment Selection")
  if null elems then error "Appointments page didn't load" else return ()

checkAppointmentsExistOnPage = do
  waitAppointmentsPageLoaded
  -- Seems that waitUntil gets affected by setImplicitWait
  -- so waitUntil 0 $ findElems ... doesn't work below
  setImplicitWait 0
  elems <- findElems (ByXPath "//button[@data-cy='inventory-button']")
  setImplicitWait defaultWait
  return . not . null $ elems

checkProvidersExistOnPage = do
  waitAppointmentsPageLoaded
  setImplicitWait 0
  elems <- findProviders
  setImplicitWait defaultWait
  return . not . null $ elems

findProviders = findElems (ByXPath "//om-provider-inventory")

readProviders =
  mapM toProvider =<< findProviders

toProvider elem = do
  location <- getText =<< findElemFrom elem (ByClass "office-name")
  let info = ProviderInfo location
  appts <- fmap keepJusts . mapM (toAppointment info)
           =<< findAppointmentsOfProvider elem
  return $ Provider info appts

findAppointmentsOfProvider elem =
  findElemsFrom elem (ByXPath ".//button[@data-cy='inventory-button']")

keepJusts = fromJust . sequence . filter isJust

toAppointment :: ProviderInfo -> Element -> WD (Maybe Appointment)
toAppointment info elem = do
  timeMaybe <- apptTime elem
  case timeMaybe of
    Nothing -> do
      logWD $ "Failed to parse appointment time"
      return Nothing
    Just time ->
      return . Just $ Appointment info time elem

apptTime elem = do
  textMaybe <- attr elem "aria-label"
  return (parseApptTimeLabel =<< textMaybe)

-- parseApptTimeLabel "Tue 01 25 3 15 pm"
parseApptTimeLabel :: Text -> Maybe LocalTime
parseApptTimeLabel
  = parseTimeM True defaultTimeLocale "%Y %a %m %d %l %M %P"
  -- Hack to add year to the current date
  . ("2022 " ++)
  . T.unpack

backToAppointmentSelection :: WD ()
backToAppointmentSelection
  = click =<< findElem (ByPartialLinkText "Back to Appointment Selection")

waitForAppointments
  :: Text -> Day -> (Appointment -> Bool) -> ChooseAppointmentPage
  -> WD [Appointment]
waitForAppointments appointmentType lastDay p page = do
  selectAppointment appointmentType page
  appts <- findAppointments lastDay p
  if null appts
    then do
      backToAppointmentSelection
      waitForAppointments appointmentType lastDay p page
    else
      return appts

selectItem :: Element -> Text -> WD [Element]
selectItem selectElem value = do
  opts <- findElemsFrom selectElem (ByTag "option")
  filterM (fmap (== value) . getText) $ opts

returnSession config wd = runSession config $ do
  sess <- getSession
  catch (wd >>= \a -> return (sess, Right a))
    $ \(e::SomeException) -> return (sess, Left e)

data FindAppointmentParams
  = FindAppointmentParams
  { user            :: Text
  , password        :: Text
  -- Choice in the "I'm looking for appointments in..." field, i.e:
  -- "SF Bay Area", "Chicago", "Ney York", etc
  , area            :: Text
  , appointmentType :: Text
  -- We search appointments from today until the `lastDay`
  , lastDay         :: Day
  -- Predicate that tells is the given appointment matches our search
  , predicate       :: Appointment -> Bool
  } -- Do not derive Show to prevent printing `user` and `password`
findAppointmentsScenario params = do
  setImplicitWait defaultWait
  appts <- login (user params) (password params)
           >>= gotoAppointmentSelection (area params)
           >>= waitForAppointments
               (appointmentType params)
               (lastDay params)
               (predicate params)
  logWD
    $ "Found matching appointments:\n"
    ++ (LT.unpack . pShowNoColor $ appts)
  return appts

nextDayStartTime day
  = LocalTime (addDays 1 day ) (TimeOfDay 0 0 0)

dayEndP m d = (<= dayEnd) . time
  where dayEnd = nextDayStartTime $ fromGregorian 2022 m d

dayStartP m d
  = (>= LocalTime (fromGregorian 2022 m d) (TimeOfDay 0 0 0))
  . time

locationsInP locations = locationsP (`elem` locations)
locationsP p
  = p . location
  . (providerInfo :: Appointment -> ProviderInfo)

timeStartP h m = (>= TimeOfDay h m 0)

timeEndP h m = (<= TimeOfDay h m 0)

allP :: [a -> Bool] -> a -> Bool
allP ps a =
  and . flip map ps $ \p -> p a

findConfirmAppointmentButton = findElem (ByXPath "//button[text() = 'Confirm Appointment']")
findOkButton = findElem (ByXPath "//button[text() = ' OK ']")

checkYourAppointmentIsBookedText = do
  elems <- findElems (ByXPath "//h1[text() = 'Your appointment is booked!']")
  if null elems
    then return False
    else return True

findPleaseTryYourSearchAgainText :: WD Element
findPleaseTryYourSearchAgainText
  = findElem (ByXPath "//h2[text() = 'Please try your search again.']")

bookAppointment :: [Appointment] -> WD (Maybe Appointment)
bookAppointment [] = return Nothing
bookAppointment (appt:appts) = do
  click . element $ appt
  click =<< findConfirmAppointmentButton
  booked <- checkYourAppointmentIsBookedText
  if booked
    then return . Just $ appt
    else do
      logWD $ "Failed to book appointment: " ++ show appt
      click =<< findOkButton
      logWD $ "Returned to appointments page"
      bookAppointment appts

bookAppointmentScenario params = do
  appts <- findAppointmentsScenario params
  apptMb <- bookAppointment appts
  case apptMb of
    Just appt -> do
      logWD $ "Booked appointment: " ++ show appt
      return appt
    Nothing -> bookAppointmentScenario params


one_medical_south_bay.hs
{-# LANGUAGE OverloadedStrings     #-}

import           Data.Text      (Text)
import qualified Data.Text      as T
import           Data.Time      (fromGregorian)
import           OneMedical
import           Test.WebDriver (runSession)

southBayLocations =
  [ "Sunnyvale"
  , "Sunnyvale Outdoor Testing Site"
  , "San Mateo - Bay Meadows"
  , "San Jose - The Alameda"
  , "San Jose - North First"
  , "Redwood City"
  , "Phillips Brooks School COVID Testing Site" -- Menlo Park
  , "Palo Alto"
  ]

params
  = FindAppointmentParams
  { user = "<your email registered with OneMedical>"
  , password = "<your OneMedical password>"
  , area = "SF Bay Area"
  , appointmentType = "COVID-19 PCR Test"
  , lastDay = fromGregorian 2022 1 11
  , predicate = locationsInP southBayLocations
  --  predicate = not . locationsInP ["UCSF One Medical Testing Site"]
  }

findSouthBayLocations = findAppointmentsScenario params
bookSouthBayLocation = bookAppointmentsScenario params
-- sess <- returnSession remoteConfig findSouthBayLocations

main = runSession remoteConfig findSouthBayLocations



app/one_medical.hs
{-# LANGUAGE OverloadedStrings     #-}

import           Data.Text      (Text)
import qualified Data.Text      as T
import           Data.Time
import           OneMedical
import           Test.WebDriver (runSession)

southBayLocations =
  [ "Sunnyvale"
  , "Sunnyvale Outdoor Testing Site"
  , "San Mateo - Bay Meadows"
  , "San Jose - The Alameda"
  , "San Jose - North First"
  , "Redwood City"
  , "Phillips Brooks School COVID Testing Site" -- Menlo Park
  , "Palo Alto"
  ]

-- Didn't check if these are all locations
sanFranciscoLocations =
  [ "Castro"
  , "Cow Hollow - Union Street"
  , "Duboce Triangle"
  , "Financial District - Clay"
  , "Financial District - Montgomery"
  , "Financial District - Sutter"
  , "Four Embarcadero Center"
  , "Hayes Valley"
  , "Inner Sunset"
  , "Laurel Heights"
  , "Mission - Valencia"
  , "Mission Bay"
  , "Noe Valley - 24th St"
  , "Pacific Heights"
  , "SoMa - South Park"
  , "Spear Street"
  , "Transbay Center"
  , "Two Embarcadero Center"
  , "UCSF One Medical Testing Site"
  , "West Portal"
  ]

params
  = FindAppointmentParams
  { user = "<your email registered with OneMedical>"
  , password = "<your OneMedical password>"
  , area = "SF Bay Area"
  , appointmentType = "COVID-19 PCR Test"
  , lastDay = fromGregorian 2022 01 30
  , predicate = locationsInP southBayLocations
  -- , predicate = not . locationsInP [ "UCSF One Medical Testing Site"
  --                                  , "Sunnyvale Outdoor Testing Site"
  --                                  ]
  -- , predicate = \appt ->
  --     not ("San Jose" `T.isInfixOf` location (providerInfo appt))
  --     && (localTimeOfDay (time appt) >= TimeOfDay 10 0 0)
  }

findAppointments = findAppointmentsScenario params
bookAppointment = bookAppointmentScenario params
-- sess <- returnSession remoteConfig findAppointments

main = runSession remoteConfig findAppointments

